src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -9731,93 +9560,302 @@ protected override bool Join(ref LocalState self, ref LocalState other)
                 }
-                set
+            }
+
+            private void SetValue(int id, int index, NullableFlowState value)
+            {
+                if (Id != id)
+                {
+                    Debug.Assert(Id > id);
+                    _container!.Value.SetValue(id, index, value);
+                }
+                else
                 {
                     // No states should be modified in unreachable code, as there is only one unreachable state.
                     if (!this.Reachable) return;
-                    slot *= 2;
-                    _state[slot] = (value != NullableFlowState.NotNull);
-                    _state[slot + 1] = (value == NullableFlowState.MaybeDefault);
+                    index *= 2;
+                    _state[index] = (value != NullableFlowState.NotNull);
+                    _state[index + 1] = (value == NullableFlowState.MaybeDefault);
+                }
+            }
+
+            internal void ForEach<TArg>(Action<int, TArg> action, TArg arg)
+            {
+                _container?.Value.ForEach(action, arg);
+                for (int index = 1; index < Capacity; index++)
+                {
+                    action(Variables.ConstructSlot(Id, index), arg);
+                }
+            }
+
+            internal LocalState GetStateForVariables(int id)
+            {
+                var state = this;
+                while (state.Id != id)
+                {
+                    state = state._container!.Value;
                 }
+                return state;
             }
 
             /// <summary>
             /// Produce a duplicate of this flow analysis state.
             /// </summary>
             /// <returns></returns>
-            public LocalState Clone() => new LocalState(_state.Clone());
+            public LocalState Clone()
+            {
+                var container = _container is null ? null : new Boxed(_container.Value.Clone());
+                return new LocalState(Id, container, _state.Clone());
+            }
 
-            public bool Join(in LocalState other) => _state.UnionWith(in other._state);
+            public bool Join(in LocalState other)
+            {
+                Debug.Assert(Id == other.Id);
+                bool result = false;
+                if (_container is { } && _container.Value.Join(in other._container!.Value))
+                {
+                    result = true;
+                }
+                if (_state.UnionWith(in other._state))
+                {
+                    result = true;
+                }
+                return result;
+            }
 
-            public bool Meet(in LocalState other) => _state.IntersectWith(in other._state);
+            public bool Meet(in LocalState other)
+            {
+                Debug.Assert(Id == other.Id);
+                bool result = false;
+                if (_container is { } && _container.Value.Meet(in other._container!.Value))
```

Comment:
>{ } [](start = 34, length = 3)

Consider using `not null`

Reply:
I think I prefer this syntax since it's shorter.

---
In reply to: [560564088](https://github.com/dotnet/roslyn/pull/50417#discussion_r560564088) [](ancestors = 560564088)